src/main/haskell/language-kore/src/Data/Kore/Unification/UnifierImpl.hs

```diff
@@ -0,0 +1,386 @@
+                , unificationSolutionConstraints = [ (v, functionalPattern) ]
+                }
+            , Proposition_5_24_3 functionalProof v functionalPattern
+            )
+        _ -> Left NonFunctionalPattern
+
+-- returns from the recursion, building the unified term and
+-- pushing up the constraints (substitution)
+postTransform
+    :: Pattern level Variable
+        (Either
+            (UnificationError level)
+            (UnificationSolution level, UnificationProof level)
+        )
+    -> Either
+        (UnificationError level)
+        (UnificationSolution level, UnificationProof level)
+postTransform (ApplicationPattern ap) = do
+    -- traceM (show ap)
+    children <- sequenceA (applicationChildren ap)
+    let (subSolutions, subProofs) = unzip children
+    return
+        ( UnificationSolution
+            { unificationSolutionTerm = Fix $ ApplicationPattern ap
+                { applicationChildren =
+                        map unificationSolutionTerm subSolutions
+                }
+            , unificationSolutionConstraints =
+                sortBy (compare `on` fst)
+                    (concatMap
+                        unificationSolutionConstraints
+                        subSolutions
+                    )
+            }
+        , AndDistributionAndConstraintLifting
+            (applicationSymbolOrAlias ap)
+            subProofs
+        )
+
+-- finds ocurrences of x = t1 /\ x = t2 and transforms them
+-- into x = (t1 /\ t2)
+normalizeSubstitution1
+    :: UnificationSubstitution level
+    -> [( Variable level
+        , Either
+            (CommonPurePattern level, UnificationProof level)
+            (CommonPurePattern level)
+        )
+       ]
+normalizeSubstitution1 ((x, t1): rest@((y, t2): subst))
+    | x == y =
+        ( x
+        , Left
+            ( Fix $ AndPattern And
+                { andSort = variableSort x
+                , andFirst = t1
+                , andSecond = t2
+                }
+            , SubstitutionMerge x t1 t2
+            )
+        ) : normalizeSubstitution1 subst
+    | otherwise =
+        ( x, Right t1) : normalizeSubstitution1 rest
+normalizeSubstitution1 [(x, t)] = [(x, Right t)]
+normalizeSubstitution1 [] = []
+
+-- solves x = (t1 /\ t2), producing x = t /\ subst
+solveEqualityConstraint
+    :: MetadataTools level
+    -> (Variable level, CommonPurePattern level, UnificationProof level)
+    -> Either
+        (UnificationError level)
+        ( UnificationSubstitution level
+        , UnificationProof level
+        )
+solveEqualityConstraint tools (v, p, ecProof) = do
+    (solution, proof) <- simplifyAnd tools p
+    return
+        ( (v, unificationSolutionTerm solution)
+          : unificationSolutionConstraints solution
+        , CombinedUnificationProof [ecProof, proof]
+        )
+
+-- iteratively finds ocurrences of x = t1 /\ x = t2, solves them, and
+-- computes the new substitution
```

Comment:
Could you explain (as a code comment) why you can't just split the substitution by variable, then foldr with a combination of normalizeSubstitution1 and solveEqualityConstraint? If you don't already have a test for this case, could you add one?